### PR TITLE
[compile] add sequence parallelism for moe

### LINKED ISSUE #29139: [Feature]: Optimize collectives in TP MoE case using torch.compile pass
### 🚀 The feature, motivation and pitch

To avoid redundant work in MoE models in the TP case, sequence parallelism was added to the Deepseek model definition in #24134 and expanded to other models in #24982. However, to avoid performing surgery on the linear layer, the current approach performs more communication than necessary. With a torch.compile custom pass, we can rewrite the graph to remove the redundant computation.

### More details

Before the SP optimization, the ops in the model were:
```
- o_proj:[num_tokens, ...] -> [num_tokens, ...] (incomplete results)
- all_reduce:[num_tokens, ...] -> [num_tokens, ...]
- rms_norm:[num_tokens, ...] -> [num_tokens, ...]
- router:[num_tokens, ...] -> [num_tokens, ...]
- experts:[num_tokens, ...] -> [num_tokens, ...]
- ...
```

With sequence parallel enabled, this becomes:
```
- o_proj: [num_tokens, ...] -> [num_tokens, ...] (incomplete results)
- all_reduce: [num_tokens, ...] -> [num_tokens, ...]
- rms_norm:[num_tokens, ...] -> [num_tokens, ...]
- chunk: [num_tokens, ...] -> [num_tokens/tp, ...]
- router: [num_tokens/tp, ...] -> [num_tokens/tp, ...]
- experts: [num_tokens/tp, ...] -> [num_tokens/tp, ...]
- all_gather: [num_tokens/tp, ...] -> [num_tokens, ...]
```

Additionally, experts now properly do the dp+tp<->ep dispatch instead of just the original replicated dp<->ep dispatch.

Notice that the `all_reduce` (and `rms_norm`) do redundant communication as each TP rank only requires partial results. With a compile pass, we can convert the `all_reduce` -> `rms_norm` -> `chunk` sequence into a `reduce_scatter` -> `rms_norm`:

```
- o_proj: [num_tokens, ...] -> [num_tokens, ...] (incomplete results)
- reduce_scatter: [num_tokens, ...] -> [num_tokens/tp, ...]
- router: [num_tokens/tp, ...] -> [num_tokens/tp, ...]
- experts: [num_tokens/tp, ...] -> [num_tokens/tp, ...]
- all_gather: [num_tokens/tp, ...] -> [num_tokens, ...]
```

We should create a new `SequenceParallelismMoEPass`, controlled by a new `PassConfig.enable_sp_moe` flag (following the new naming convention in #27995) so that it can be turned on independently of regular SP. We will likely need to pad the number of tokens to a multiple of TP size, although like described in #29136, there are alternatives.

#### SP
Additionally, to make SP-MoE compatible with AsyncTP, the SP pass should also replace the `all_gather` -> `rms_norm` sequence with `rms_norm` -> `all_gather`, such that the `all_gather` can be fused with the following QKV proj. We can see that in the SP case we perform an `all_gather` instead of `all_reduce` after the MLP layer.

### Alternatives

Alternatively, the original optimization could be done as a compile pass as well, which would significantly clean up the MoE model definitions. However, that would mean that `VLLM_COMPILE` compilation mode would be required for this optimization and if compilation is disabled, the optimization would be disabled as well. Generally we accept lower performance in eager mode as compilation is on by default, but I know there was a reason this was done this way (don't remember why).

We could also just let the SP pass first transform `all_reduce` -> `rms_norm` into `reduce_scatter` -> `rms_norm` -> `all_gather` and then eliminate the `all_gather` -> `chunk` combo. But then we have less control over SP after MoE (although it seems like all_reduce is not even used in that case).

### Additional context

Original proposal comment: https://github.com/vllm-project/vllm/pull/24982#pullrequestreview-3259494618

cc @tlrmchlsmth @bnellnm @robertgshaw2-redhat @alexm-redhat @zou3519 @nvpohanh @youkaichao 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by anant37289**
## Purpose
Tranform a sequence of `allreduce->chunk` to `reduce_scatter` for moe models with `sequence_parallel_chunk` ops. Resolves #29139 
## Test Plan
`pytest -v tests/compile/distributed/test_sequence_parallelism_moe.py`
## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/compilation/sequence_parallelism_moe.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            seq_len = input.size(0)
+            remainder = seq_len % self.tp_size
+            if remainder != 0:
+                pad_len = self.tp_size - remainder
+                y = nn.functional.pad(input, (0, 0, 0, pad_len))
+            else:
+                y = input
+            reduce_scatter = self._reduce_scatter(y)
+
+            return reduce_scatter
```
Comment (ProExpertProg) [inline vllm/compilation/sequence_parallelism_moe.py]: Try this:

```
pad_len = (-input.size(0)) % self.tp_size
y = nn.functional.pad(input, (0, 0, 0, pad_len))
reduce_scatter = self._reduce_scatter(y)
```

File: vllm/config/compilation.py [hunk 2]
```diff
@@ -107,6 +107,8 @@ class PassConfig:
     """Whether to enable the custom no-op elimination pass."""
     enable_sequence_parallelism: bool = False
     """Whether to enable sequence parallelism."""
+    enable_sp_moe: bool = False
```
Comment (ZJY0516) [inline vllm/config/compilation.py]: Could we unify `enable_sp_moe` and `enable_sequence_parallelism` into a single `enable_sp` configuration? @ProExpertProg

Comment (ProExpertProg) [reply vllm/config/compilation.py]: I think no because MoE models have some dense layers as well, so the two flags do different things. With optimization levels that just landed, we will enable both by default anyway so having separate flags is better in my opinion to give users more control and for easier debugging & perf impact measurements.

**Review by ProExpertProg — COMMENTED** [review]
Looks good overall! Can you:
1. Add a test case to `test_e2e_fusions.py`
2. Report lm_eval numbers?
3. Report performance numbers?

For perf and accuracy evals, I would look at the following models:
- Deepseek R1
- Qwen 3 30B
- Llama4

As well as the following deployment scenarios:
- TP=4
- DP=2 TP=2 `--enable-expert-parallelism`
- DP=2 TP=4 `--enable-expert-parallelism`
- DP=4 TP=2 `--enable-expert-parallelism`

**Comment by anant37289** [comment]
Hi so i actually did not have the compute beyond 2 24 GB gpus to do this, so the earlier small models ran file but it was not able to support even the Qwen3-30b-A8b model. I tried renting 2xA40s, but I am running into a couple issues:
1. The sequence parallel chunking from earlier is only applied when TP size > 1 and DP size > 1, I do not see the reason for them both to be atleast 2. Am i missing something?
```
    @property
    def use_sequence_parallel_moe(self) -> bool:
        return (
            self.all2all_backend
            in (
                "allgather_reducescatter",
                "naive",
                "deepep_high_throughput",
                "deepep_low_latency",
            )
            and self.enable_expert_parallel
            and self.tensor_parallel_size > 1
            and self.data_parallel_size > 1
        )
```

2. I tried to make it work with TP=2 and DP=1. but it errors out, probably I am wrong in doing this(I am yet to test on TP=2, DP=2), so i wanted to confirm this once.

**Comment by ProExpertProg** [comment]
Okay, sounds good, please test whatever you can on the hardware you have. Once we're ready to merge I can help run some tests (I am away for 2 weeks so will be after that).

I didn't realize we didn't do MoE SP unless DP>1 (TP>1 makes sense). Feel free to only test TP>1 & DP>1 in that case.

**Comment by anant37289** [comment]
Hi @ProExpertProg 
> I didn't realize we didn't do MoE SP unless DP>1 (TP>1 makes sense). Feel free to only test TP>1 & DP>1 in that case.

About this. Is there a reason for doing this? After merging from main I was able to get sequence parallel chunking from earlier to work with TP=2 and DP=1(by just changing the above mention `use_sequence_parallel_moe` function(checking only for TP).
